File: 射击类小游戏/射击类小游戏/ViewController.m
//
//  ViewController.m
//  射击类小游戏
//
//  Created by LJP on 6/12/17.
//  Copyright © 2017年 poco. All rights reserved.
//

#import "ViewController.h"

#import "Ship.h"
#import "Bullet.h"

@interface ViewController () <ARSCNViewDelegate,SCNPhysicsContactDelegate>

@property(nonatomic, strong) UILabel * l;

//视图
@property(nonatomic, strong) ARSCNView * jpARSCNView;

//会话
@property(nonatomic, strong) ARSession * jpARSession;

//跟踪会话
@property(nonatomic, strong) ARWorldTrackingConfiguration * jpARWTkConfiguration;


@end

    
@implementation ViewController

- (void)viewDidLoad {
    
    [super viewDidLoad];

    [self.view addSubview:self.jpARSCNView];
    
    [self initUI];
    
    [self addNewShip];
    
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    [self.jpARSCNView.session runWithConfiguration:self.jpARWTkConfiguration];
    
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];

}


#pragma mark - 初始化UI

-(void)initUI {
    
    UILabel * l = [[UILabel alloc]init];
    
    l.frame = CGRectMake(self.view.frame.size.width/2-50, 60, 100, 60);
    
    l.text = @"0";
    
    l.textColor = [UIColor whiteColor];
    
    l.textAlignment = NSTextAlignmentCenter;
    
    self.l = l;
    
    [self.view addSubview:l];
}

#pragma mark - 私有方法


- (void)addNewShip {
    
    Ship * cubeNode = [[Ship alloc]init];
    
    float x = [self floatBetween];
    float y = [self floatBetween];
    
    cubeNode.position = SCNVector3Make(x, y, -1);
    
    [self.jpARSCNView.scene.rootNode addChildNode:cubeNode];
    
}


//生产随机数
- (CGFloat )floatBetween {

    CGFloat a = (float)(1+arc4random()%99)/100-0.5;
    
    NSLog(@"a == %f",a);
    
    return a;
    
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {

    NSLog(@"点击了");


    Bullet * bulletsNode = [[Bullet alloc]init];

    NSArray * arr = [self getUserVector];

    NSValue * dirValue = arr[0];
    NSValue * posValue = arr[1];

    SCNVector3 direction = [dirValue SCNVector3Value];
    SCNVector3 position  = [posValue SCNVector3Value];

    bulletsNode.position = position;

    SCNVector3 bulletDirection = direction;

    [bulletsNode.physicsBody applyForce:bulletDirection impulse:YES];

    [self.jpARSCNView.scene.rootNode addChildNode:bulletsNode];

}

- (void)removeNodeWithAnimation:(SCNNode *)node explosion:(BOOL)explosion {
    
    NSLog(@"动画");
    
    if (explosion) {
        
        SCNParticleSystem * particleSystem = [SCNParticleSystem particleSystemNamed:@"explosion" inDirectory:nil];
        
        SCNNode * systemNode = [[SCNNode alloc]init];
        
        [systemNode addParticleSystem:particleSystem];
        
        systemNode.position = node.position;
        
        [self.jpARSCNView.scene.rootNode addChildNode:systemNode];
        
    }
    
    [node removeFromParentNode];
    
}

- (NSArray *)getUserVector {

    ARFrame * frame = self.jpARSCNView.session.currentFrame;
    
    if (frame) {
        
        SCNMatrix4 mat = SCNMatrix4FromMat4(frame.camera.transform);
        
        SCNVector3 dir = SCNVector3Make(-1* mat.m31, -1* mat.m32 , -1* mat.m33);
        
        SCNVector3 pos = SCNVector3Make(mat.m41,mat.m42 ,mat.m43);
        
        NSValue * dirValue = [NSValue valueWithSCNVector3:dir];
        NSValue * posValue = [NSValue valueWithSCNVector3:pos];
        
        return @[dirValue,posValue];
        
    }else {
        
        NSValue * dirValue = [NSValue valueWithSCNVector3:(SCNVector3Make(0, 0, -1))];
        NSValue * posValue = [NSValue valueWithSCNVector3:(SCNVector3Make(0, 0, -0.2))];
        
        return @[dirValue,posValue];
    }
    
}


#pragma mark - ARSCNViewDelegate

- (void)session:(ARSession *)session didFailWithError:(NSError *)error {
    // Present an error message to the user
    
}

- (void)sessionWasInterrupted:(ARSession *)session {
    // Inform the user that the session has been interrupted, for example, by presenting an overlay
    
}

- (void)sessionInterruptionEnded:(ARSession *)session {
    // Reset tracking and/or remove existing anchors if consistent tracking is required
    
}

// 这里是自己独立开一条线程出来
- (void)physicsWorld:(SCNPhysicsWorld *)world didBeginContact:(SCNPhysicsContact *)contact {
    
    NSLog(@"触发");
    
    if (contact.nodeA.physicsBody.categoryBitMask == 2 ||
        contact.nodeB.physicsBody.categoryBitMask == 2) {
        
        [self removeNodeWithAnimation:contact.nodeB explosion:NO];
        
        dispatch_sync(dispatch_get_main_queue(), ^(){
            self.l.text = [NSString stringWithFormat:@"%ld", ([self.l.text integerValue]+1)];
        });
        
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            
            [self removeNodeWithAnimation:contact.nodeA explosion:YES];
            [self addNewShip];
            
        });
        
    }
    
}



#pragma mark - 访问器方法

- (ARSCNView *)jpARSCNView {
    
    if (_jpARSCNView == nil) {
        
        _jpARSCNView = [[ARSCNView alloc]init];
        
        _jpARSCNView.frame = self.view.frame;
        
        _jpARSCNView.delegate = self;
        
        _jpARSCNView.showsStatistics = YES;
        
        SCNScene * scene= [[SCNScene alloc]init];
        _jpARSCNView.scene = scene;
        
        _jpARSCNView.scene.physicsWorld.contactDelegate = self;
        
    }
    
    return _jpARSCNView;
}

- (ARSession *)jpARSession {
    
    if (_jpARSession == nil) {
        _jpARSession = [[ARSession alloc]init];
    }
    
    return _jpARSession;
    
}

- (ARWorldTrackingConfiguration *)jpARWTkConfiguration {
    
    if (_jpARWTkConfiguration == nil) {
        _jpARWTkConfiguration = [[ARWorldTrackingConfiguration alloc]init];
        _jpARWTkConfiguration.planeDetection = ARPlaneDetectionHorizontal;
//        _jpARWTkConfiguration.lightEstimationEnabled = YES;
    }
    
    return _jpARWTkConfiguration;
    
}


@end

File: otherDemo/ARKitDemo-master/myARKitDemo/ViewController.swift
//
//  ViewController.swift
//  myARKitDemo
//
//  Created by ksyun on 2017/6/13.
//  Copyright © 2017年 ksyun. All rights reserved.
//

import UIKit
import SceneKit
import ARKit

class ViewController: UIViewController, ARSCNViewDelegate,SCNPhysicsContactDelegate {

    @IBOutlet var sceneView: ARSCNView!
    
    open var myPlane:plane!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.setupScene()
    }
    
    func setupScene()->Void{
        // Set the view's delegate
        sceneView.delegate = self
        
        // Show statistics such as fps and timing information
        //sceneView.scene = SCNScene(named: "art.scnassets/billboard.dae")!
        
        //let ship:SCNNode = sceneView.scene.rootNode.childNode(withName: "Billboard", recursively: true)!
        
        sceneView.showsStatistics = true
        
        sceneView.autoenablesDefaultLighting = true
        
        sceneView.automaticallyUpdatesLighting = true
        
        sceneView.debugOptions = [ARSCNDebugOptions.showWorldOrigin,ARSCNDebugOptions.showFeaturePoints]
        
        //ship and bullet
        self.addNewShip()
        
        let tapGesture = UITapGestureRecognizer(target: self, action: #selector(ViewController.handleTapGesture))
        //设置手势点击数,双击：点2下
        tapGesture.numberOfTapsRequired = 1
        self.view.addGestureRecognizer(tapGesture)
        
        sceneView.scene.physicsWorld.contactDelegate = self
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        self.setupSession()
    }
    
    func setupSession(){
        // Create a session configuration
        let configuration = ARWorldTrackingConfiguration()
        configuration.planeDetection = .horizontal
        
        // Run the view's session
        sceneView.session.run(configuration)
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        
        // Pause the view's session
        sceneView.session.pause()
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Release any cached data, images, etc that aren't in use.
    }
    
    //MARK: - SHIP AND BULLTE
    func addNewShip() {
        let cubeNode = Ship()
        
        let posX = floatBetween(-0.5, and: 0.5)
        let posY = floatBetween(-0.5, and: 0.5  )
        cubeNode.position = SCNVector3(posX, posY, -1) // SceneKit/AR coordinates are in meters
        sceneView.scene.rootNode.addChildNode(cubeNode)
    }
    
    @objc func handleTapGesture(sender: UITapGestureRecognizer){
        let bulletsNode = Bullet()
        
        let (direction, position) = self.getUserVector()
        bulletsNode.position = position // SceneKit/AR coordinates are in meters
        
        let bulletDirection = direction
        bulletsNode.physicsBody?.applyForce(bulletDirection, asImpulse: true)
        sceneView.scene.rootNode.addChildNode(bulletsNode)
    }
    
    func removeNodeWithAnimation(_ node: SCNNode, explosion: Bool) {
        if explosion {
            let particleSystem = SCNParticleSystem(named: "explosion", inDirectory: nil)
            let systemNode = SCNNode()
            systemNode.addParticleSystem(particleSystem!)
            // place explosion where node is
            systemNode.position = node.position
            sceneView.scene.rootNode.addChildNode(systemNode)
        }
        
        // remove node
        node.removeFromParentNode()
    }
    
    func getUserVector() -> (SCNVector3, SCNVector3) { // (direction, position)
        if let frame = self.sceneView.session.currentFrame {
            let mat = SCNMatrix4.init(frame.camera.transform) // 4x4 transform matrix describing camera in world space
            let dir = SCNVector3(-1 * mat.m31, -1 * mat.m32, -1 * mat.m33) // orientation of camera in world space
            let pos = SCNVector3(mat.m41, mat.m42, mat.m43) // location of camera in world space
            
            return (dir, pos)
        }
        return (SCNVector3(0, 0, -1), SCNVector3(0, 0, -0.2))
    }
    
    func floatBetween(_ first: Float,  and second: Float) -> Float { // random float between upper and lower bound (inclusive)
        return (Float(arc4random()) / Float(UInt32.max)) * (first - second) + second
    }

    // MARK: - ARSCNViewDelegate
    
    //添加节点时候调用（当开启平地捕捉模式之后，如果捕捉到平地，ARKit会自动添加一个平地节点）
    func renderer(_ renderer: SCNSceneRenderer, didAdd node: SCNNode, for anchor: ARAnchor) {
        if anchor.isMember(of: ARPlaneAnchor.self){
            myPlane = plane(anchor: anchor as! ARPlaneAnchor)
            node.addChildNode(myPlane)
        }
    }
    
    func renderer(_ renderer: SCNSceneRenderer, didUpdate node: SCNNode, for anchor: ARAnchor) {
        myPlane.update(anchor: anchor as! ARPlaneAnchor)
    }
    
    func session(_ session: ARSession, didFailWithError error: Error) {
        // Present an error message to the user
        
    }
    
    func sessionWasInterrupted(_ session: ARSession) {
        // Inform the user that the session has been interrupted, for example, by presenting an overlay
        
    }
    
    func sessionInterruptionEnded(_ session: ARSession) {
        // Reset tracking and/or remove existing anchors if consistent tracking is required
        
    }
    
    //MARK: - SCNPhysicsContactDelegate
    func physicsWorld(_ world: SCNPhysicsWorld, didBegin contact: SCNPhysicsContact) {
        //print("did begin contact", contact.nodeA.physicsBody!.categoryBitMask, contact.nodeB.physicsBody!.categoryBitMask)
        if contact.nodeA.physicsBody?.categoryBitMask == CollisionCategory.ship.rawValue || contact.nodeB.physicsBody?.categoryBitMask == CollisionCategory.ship.rawValue { // this conditional is not required--we've used the bit masks to ensure only one type of collision takes place--will be necessary as soon as more collisions are created/enabled
            print("Hit ship!")
            self.removeNodeWithAnimation(contact.nodeB, explosion: false) // remove the bullet
            
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.5, execute: { // remove/replace ship after half a second to visualize collision
                self.removeNodeWithAnimation(contact.nodeA, explosion: true)
                self.addNewShip()
            })
            
        }
    }
}

struct CollisionCategory: OptionSet {
    let rawValue: Int
    
    static let bullets  = CollisionCategory(rawValue: 1 << 0) // 00...01
    static let ship = CollisionCategory(rawValue: 1 << 1) // 00..10
}
